### PR TITLE
Add streaming tree clone and deep-link invitation support

### PR DESCRIPTION
**Pull request by kumavis**
## Description

This PR implements two major features.

### 1. Streaming Tree Clone (`@endo/endo-fs`)
Adds a high-performance, memory-bounded mechanism to ship entire directory trees as a single ordered frame stream rather than a client-driven pipelined walk. This implements Pillar 3c of `designs/endo-app-sharing.md`.

**Key files:**
- `packages/endo-fs/src/clone.js` — Core streaming clone implementation with:
  - `streamTree(sourceRoot, options)` — Serializes a source tree into a `PassableReader`
  - `writeTreeStream(destRoot, reader)` — Drains the stream into a destination `Directory`
  - `cloneTree(sourceRoot, destRoot, options)` — Convenience wrapper combining both
  - Depth-first traversal with stable name-sorted ordering for reproducibility
  - Large files streamed as base64-encoded chunks within the single frame stream (no whole-file buffering)
  - Pattern-validated frames via `CloneFrameShape` on both the producer (`streamTree`) and consumer (`writeTreeStream`) sides
  - Destination files created with `{ truncate: true }` so re-cloning over a longer pre-existing file leaves no stale trailing bytes

- `packages/endo-fs/test/clone.test.js` — Test suite covering full tree round-trip, empty files/directories, multi-chunk large files, sub-directory cloning, explicit `streamTree` + `writeTreeStream` composition, the empty-tree edge case, and overwrite-into-a-non-empty-destination.

- `packages/endo-fs/src/index.js` — Exports the clone API

**Design decisions realized:**
- One stream, not pipelined walks — eliminates round-trip latency
- No content hashing — integrity is the transport's job (secure channel)
- Large files stream as chunks within the single stream
- Pluggable durable destination via ordinary `Directory` cap
- Base64 encoding for CapTP marshalling compatibility

### 2. Deep-Link Invitation Support (Familiar shell + Chat)
Implements an `endo://` protocol handler for peer invitations, so users can accept (and share) peers by clicking links. This implements `designs/familiar-deep-link-invitations.md`.

**Format.** The deep link is an **intent-path** URL, distinct from the daemon's internal locator:

```
endo://invite?node={node}&id={inv}&from={handle}[&fromNode={n}]&at={addr}...
```

The authority segment names the *intent* (`invite`) so `endo://` can route future actions without overloading a locator's `type`. The link carries exactly the fields `host.accept` consumes — `node`, `id`, the required inviter handle `from`, an optional `fromNode` (agent-key case), and `at` connection hints. The handler validates the link and **reconstructs** the canonical daemon locator (`endo://{node}/?id=…&type=invitation&from=…&at=…`) that `host.accept` takes; the link is not passed verbatim.

**Key files:**
- `packages/familiar/src/deep-link.js` — Pure, dependency-free (SES-free, for the Electron main process):
  - `parseInviteUrl(text)` — validates an `endo://invite?…` link and returns `{ locator, node, number, from, fromNode, fingerprint, addresses }` (locator = reconstructed canonical form)
  - `formatInviteUrl(parts)` — inverse of `parseInviteUrl`, so producer and receiver share one contract
  - `isInviteUrl(text)` / `findInviteUrlInArgv(argv)` — boolean check and Windows/Linux cold-start argv extraction
- `packages/familiar/electron-main.js` — Registers `endo://` as protocol client; single-instance lock; `open-url` (macOS) and `second-instance` (Windows/Linux) handlers; queues invites until the renderer is ready; IPC handlers registered before window creation to avoid a cold-start race.
- `packages/familiar/preload.js` — Bridge: `getPendingDeepLinkInvite()` and `onDeepLinkInvite(callback)` (renderer subscribes first, then pulls queued invites).
- `packages/chat/deep-link-invite.js` — Routes a captured invite into the existing `accept` command form (the form is the confirmation screen).
- `packages/chat/invite-link.js` — Producer/receiver glue in the renderer: `/invite` surfaces the shareable `endo://invite?…` link; `/accept` takes either that link or a raw locator (`normalizeInvitationInput`). Mirrors the `deep-link.js` contract since Chat cannot import the Familiar shell.

**Follow-up (not in this PR):** an *open* (multi-use) invitation. Today `host.invite` is single-take (consume-once); the deep-link format already accommodates a reusable invite, but multi-use acceptance is new daemon work tracked separately.

https://claude.ai/code/session_0112zLHkZKDDUzUDEwPvScwT

### CODE REVIEW

File: packages/endo-fs/src/clone.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+          await E(destRoot).materialise(harden(frame.path), {});
+          directories += 1;
+          break;
+        }
+        case 'file': {
+          !writer || Fail`clone stream: nested file frame`;
+          const { parent, name } = await resolveParent(destRoot, frame.path);
+          openFile = await E(parent).create(name, {});
+          writer = iterateBytesWriter(await E(openFile).write(0n));
+          files += 1;
```
[the remark this answers is not in the document]
Comment (kumavis) [reply packages/endo-fs/src/clone.js]: Good catch — confirmed real. `Directory.create(name, {})` pwrites `EMPTY_BYTES` (which doesn't shrink an existing file) and `OpenFile.write` is pwrite with no tail truncate, so re-cloning a shorter file over a longer one left stale trailing bytes. Fixed in bcbc87eb: destination files are now created with `create(name, { truncate: true })`, plus a regression test cloning over a longer pre-existing file.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/endo-fs/src/clone.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+  /** @type {any} */
+  let openFile;
+  // The exo-stream byte writer (BytesWriterIterator) is iterator-shaped but
+  // not a full AsyncGenerator (no [Symbol.asyncDispose], 0-arg return()), so
+  // it is kept loosely typed.
+  /** @type {any} */
+  let writer;
+
+  try {
+    for await (const frame of iterateReader(reader)) {
```
[the remark this answers is not in the document]
Comment (kumavis) [reply packages/endo-fs/src/clone.js]: Done in bcbc87eb — `writeTreeStream` now drains via `iterateReader(reader, { readPattern: CloneFrameShape })`, so a malformed frame breaks at the stream boundary on the consumer side too rather than failing deep in the switch (or inside `decodeBase64`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/endo-fs/test/clone.test.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  t.is(fromUtf8(await readFileAt(dest, ['src', 'deep', 'nested.txt'])), big);
+
+  // Empty directory is materialised even though it has no children.
+  const top = await childKinds(dest);
+  t.is(top.get('emptydir'), 'directory');
+  t.is(top.get('src'), 'directory');
+  t.is(top.get('readme.txt'), 'file');
+  t.is(top.get('empty.txt'), 'file');
+});
+
```
[the remark this answers is not in the document]
Comment (kumavis) [reply packages/endo-fs/test/clone.test.js:148]: Added in bcbc87eb — `cloneTree into a non-empty destination overwrites without stale tail bytes` pre-populates the destination with a longer file, clones, and asserts the result equals the source exactly (no leftover tail). It fails on the old `create(name, {})` and passes with the truncate fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/chat/deep-link-invite.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
+// @ts-check
+/**
+ * Route `endo://` deep-link peer invitations into the Accept command form
+ * (designs/familiar-deep-link-invitations.md).
+ *
+ * When the Familiar shell captures an `endo://...&type=invitation` link, it
+ * forwards the parsed invite to this renderer over the preload bridge
```
[the remark this answers is not in the document]
Comment (kumavis) [reply packages/chat/deep-link-invite.js]: Fixed in bcbc87eb — the header now reads `endo://invite?...`; `type=invitation` only appears in the reconstructed daemon locator, as you note.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/familiar/preload.js [hunk 5]
```diff
@@ -20,5 +20,14 @@ contextBridge.exposeInMainWorld(
       ipcRenderer.on('familiar:security-warnings', (_event, warnings) =>
         callback(warnings),
       ),
+    // endo:// deep-link peer invitations
+    // (designs/familiar-deep-link-invitations.md). The renderer pulls any
+    // invite queued before it was listening, then subscribes for live ones.
```
[the remark this answers is not in the document]
Comment (kumavis) [reply packages/familiar/preload.js]: Fixed in bcbc87eb — the comment now documents the safe order (subscribe via `onDeepLinkInvite` first, then pull via `getPendingDeepLinkInvite`) and notes that subscribing first is what closes the cold-start race.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kumavis** [comment]
Closing in favour of two focused PRs, one per feature (both based on `llm`):

- **#398 — `feat(endo-fs): streaming tree clone`** — `packages/endo-fs/**` + `designs/endo-app-sharing.md`.
- **#399 — `feat(familiar,chat): endo:// deep-link peer invitations`** — `packages/chat/**` + `packages/familiar/**` + `designs/familiar-deep-link-invitations.md`.

All review feedback from this PR (Copilot + self-review) is carried over and resolved in the splits. Outstanding follow-ups are now documented in each feature's design doc (clone: § Known Gaps; deep-link: § Outstanding / Follow-ups). The milestone-level design docs (`app-sharing-milestone.md`, `familiar-app-ui-hosting.md`, `designs/README.md`) are intentionally left to the design PR (#381) rather than duplicated here.

---
_Generated by [Claude Code](https://claude.ai/code)_
